// kayn/dist/lib/Endpoints/SummonerEndpoint.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _Endpoint2 = require('../Endpoint');

var _Endpoint3 = _interopRequireDefault(_Endpoint2);

var _Request = require('../RequestClient/Request');

var _Request2 = _interopRequireDefault(_Request);

var _methodNames = require('../Enums/method-names');

var _methodNames2 = _interopRequireDefault(_methodNames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var SummonerEndpoint = function (_Endpoint) {
  _inherits(SummonerEndpoint, _Endpoint);

  function SummonerEndpoint(config, limiter) {
    _classCallCheck(this, SummonerEndpoint);

    var _this = _possibleConstructorReturn(this, (SummonerEndpoint.__proto__ || Object.getPrototypeOf(SummonerEndpoint)).call(this));

    _this.config = config;

    _this.by = {
      name: _this.summonerName.bind(_this),
      id: _this.summonerID.bind(_this),
      accountID: _this.accountID.bind(_this)
    };

    _this.serviceName = 'summoner';

    _this.limiter = limiter;
    return _this;
  }

  /**
   * Get a summoner by summoner name. 
   * 
   * Implements /lol/summoner/v3/summoners/by-name/{summonerName}.
   *  
   * @param {string} summonerName - The name of the summoner. 
   */


  _createClass(SummonerEndpoint, [{
    key: 'summonerName',
    value: function summonerName(_summonerName) {
      return new _Request2.default(this.config, this.serviceName, 'summoners/by-name/' + encodeURI(_summonerName), _methodNames2.default.SUMMONER.GET_BY_SUMMONER_NAME, 'GET', this.limiter);
    }

    /**
     * Get a summoner by summoner ID. 
     * 
     * Implements /lol/summoner/v3/summoners/{summonerId}.
     *  
     * @param {number} summonerID - The id of the summoner. 
     */

  }, {
    key: 'summonerID',
    value: function summonerID(_summonerID) {
      return new _Request2.default(this.config, this.serviceName, 'summoners/' + _summonerID, _methodNames2.default.SUMMONER.GET_BY_SUMMONER_ID, 'GET', this.limiter);
    }

    /**
     * Get a summoner by account ID. 
     * 
     * Implements /lol/summoner/v3/summoners/by-account/{accountId}.
     *  
     * @param {number} accountID - The account ID of the summoner. 
     */

  }, {
    key: 'accountID',
    value: function accountID(_accountID) {
      return new _Request2.default(this.config, this.serviceName, 'summoners/by-account/' + _accountID, _methodNames2.default.SUMMONER.GET_BY_ACCOUNT_ID, 'GET', this.limiter);
    }
  }]);

  return SummonerEndpoint;
}(_Endpoint3.default);

exports.default = SummonerEndpoint;
